src/diem/offchain/client.py

```diff
@@ -82,30 +99,44 @@ def send_request(
             raise CommandResponseError(cmd_resp)
         return cmd_resp
 
-    def process_inbound_request(self, request_sender_address: str, request_bytes: bytes) -> PaymentCommand:
+    def process_inbound_request(self, request_sender_address: str, request_bytes: bytes) -> CommandVariant:
         if not request_sender_address:
             raise protocol_error(ErrorCode.missing_http_header, f"missing {http_header.X_REQUEST_SENDER_ADDRESS}")
         try:
             _, public_key = self.get_base_url_and_compliance_key(request_sender_address)
         except ValueError as e:
             raise protocol_error(ErrorCode.invalid_x_request_sender_address, str(e)) from e
 
-        request = _deserialize_jws(request_bytes, CommandRequestObject, public_key, command_error)
-        if request.command_type == CommandType.PaymentCommand:
-            payment = typing.cast(PaymentCommandObject, request.command).payment
+        command_request_object = _deserialize_jws(request_bytes, CommandRequestObject, public_key, command_error)
+
+        if command_request_object.command_type == CommandType.PaymentCommand:
+            payment = command_request_object.command.payment
             self.validate_addresses(payment, request_sender_address)
-            cmd = self.create_inbound_payment_command(request.cid, payment)
+            cmd = self.create_inbound_payment_command(command_request_object.cid, payment)
             if cmd.is_initial():
                 self.validate_dual_attestation_limit(cmd.payment.action)
             elif cmd.is_rsend():
                 self.validate_recipient_signature(cmd, public_key)
             return cmd
+        elif command_request_object.command_type == CommandType.FundPullPreApprovalCommand:
+            fund_pull_pre_approval = command_request_object.command.fund_pull_pre_approval
+            self.validate_expiration_timestamp(fund_pull_pre_approval)
+            return self.create_inbound_funds_pull_pre_approval_command(
+                command_request_object.cid, fund_pull_pre_approval
+            )
 
         raise command_error(
             ErrorCode.unknown_command_type,
-            f"unknown command_type: {request.command_type}",
+            f"unknown command_type: {command_request_object.command_type}",
         )
 
+    def validate_expiration_timestamp(self, fund_pull_pre_approval):
+        if fund_pull_pre_approval.scope.expiration_timestamp < time.time():
```

Comment:
it is tricky to tell what is right "time", hence this validation seems not necessary, we should save the command first, then expiration check kicks in when process the command, because the machine received the request may out of sync time, but processing command can be retried with different machines

Reply:
removed